---
title: |
    | iGUIDE Summary Report
date: |
    | report generated : `r Sys.Date()`
    | software version : `r soft_version`
    | environment build : `r build_version`
    | 
    | author(s): `r signature`
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: 
      collapsed: false
    number_sections: false
    fig_caption: true
    theme: cerulean 
    highlight: tango
    smart: false
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: false
    fig_caption: true
header-includes: 
  - \usepackage{float,indentfirst,booktabs,longtable,caption}
  - \setlength{\defaultaddspace}{0em}
  - \setlength{\parindent}{2em}
  - \captionsetup{singlelinecheck=false,justification=raggedright}
  - \captionsetup{margin=25pt,labelfont=bf,labelsep=period}
fontsize: 11pt
geometry: margin=0.5in
---

```{r setup, include=FALSE}
packs <- c(
  "magrittr", "knitr", "scales", "RColorBrewer", 
  "tidyverse", "ggplot2", "ggforce", "kableExtra"
)

packs_loaded <- suppressMessages(sapply(packs, require, character.only = TRUE))

if( !all(packs_loaded) ){
  
  print(
    data.frame(
      "R-Packages" = names(packs_loaded), 
      "Loaded" = packs_loaded
    ), 
    row.names = FALSE
  )
  stop("Check dependancies.")
  
}

options(
  stringsAsFactors = FALSE, 
  scipen = 99, 
  knitr.table.format = "latex"
)

knitr::opts_chunk$set(
  echo = FALSE,
  comment = "",
  warning = FALSE,
  error = TRUE,
  message = FALSE,
  cache = FALSE,
  results = "asis",
  fig.align = "center",
  dpi = 300,
  dev = c("png", "pdf"),
  fig.pos = "H",
  out.width = "100%"
)

figure_width <- 7.5

if( args$figures != FALSE ){
  knitr::opts_chunk$set(fig.path = paste0(figure_path, "/"))
}

report_format <- ifelse(args$format == "pdf", "latex", "html")

# Helpful functions
pNums <- function(x, na.replace = 0, ...){
  
  if( is.numeric(na.replace) )x <- ifelse(is.na(x), na.replace, x)
  
  x <- format(x, big.mark = ",", ...)
  
  if( is.character(na.replace) ){
    replace_idx <- which(stringr::str_trim(x) == "NA")
    char_width <- unique(nchar(x))
    x[replace_idx] <- stringr::str_pad(
      string = rep(na.replace, length(replace_idx)), 
      width = char_width - nchar(na.replace), 
      side = "left"
    )
  }
  
  x
  
}

# Custom theme
custom_theme <- theme_bw() +
  theme(
    panel.background = element_rect(color = "black", fill = "white"),
    axis.line = element_blank(),
    axis.ticks = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(face = "bold", angle = 0),
    axis.title = element_text(color = "black", face = "bold"),
    strip.background = element_rect(color = "black", fill = "white"),
    strip.text = element_text(face = "bold"),
    panel.grid = element_blank(),
    legend.key = element_rect(fill = "white"),
    title = element_text(face = "bold"))

blank_theme <- theme_bw() +
  theme(
    panel.background = element_blank(),
    panel.border = element_blank(),
    axis.line = element_blank(),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_blank(),
    strip.background = element_blank(),
    strip.text = element_text(face = "bold"),
    panel.grid = element_blank(),
    legend.position = "none",
    title = element_text(face = "bold"))

bootstrap_options <- c("striped", "hover")
latex_options <- c("hold_position", "repeat_header")

# Captions
tbl_caps <- c(
  "Analysis summary.",
  "Specimen summary.",
  "Target sequences and associated information.",
  "Percent On-target.",
  "Off-target Loci.",
  "Cancer-associated Gene editing enrichment."
)

fig_caps <- c(
  "Genomic distribution of incorporation sites by bioinformatic characteristics.",
  "Sequence similarity between off-target sites and targeting sequence(s).",
  "Distance distribution of observed incorporation sites from On-target loci."
)

if( args$format == "html" ){
  
  tbl_nums <- if( nrow(target_tbl) == 1 ){
    tbl_nums <- 1:6
  }else{
    tbl_nums <- c(1:4,6,7)
  }
  
  tbl_caps <- paste0("Table ", tbl_nums, ". ", tbl_caps)
}

fig_caps <- paste0("**Figure ", 1:3, "**. ", fig_caps)

```

```{r title_graphic, fig.width=6, fig.height=6, out.width="50%", eval = args$graphic}
plotGenomicDensity(
  graphic_grl, res = 1E7, abund.method = "log.count", clean = TRUE
)
```

***

This report includes data from the following sequencing runs: `r gsub("_", "-", set_names)`.


```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

# Summary

The following document summarizes the results of processing `r gsub("_", "-", set_names)` sequencing set(s) through the iGUIDE pipeline. Included in this document are explanations of the data analytics as well as tables and graphics of the data obtained from the sequence analysis. This report includes `r length(unique(sample_info$specimen))` specimens treated with `r nrow(target_tbl)` targeting sequences. A total of `r pNums(round(sum(incorp_data$algnmts$count)))` reads are considered in this analysis, which represent `r pNums(round(sum(incorp_data$algnmts$contrib)))` observed incorporated double-stranded oligo-dinucleotides (dsODNs, a unit of measure associated with iGUIDE or GUIDE-seq based analyses). 

**Table 1** highlights some key information from the data analysis for each specimen, including the total number of alignments (representing the observed incorporated dsODNs), an estimated range of On-target editing efficiency, and the number of predicted off-target sites.

```{r report_summary}
eval_summary[is.na(eval_summary)] <- 0

if( args$tables ){
  
  write.csv(
    x = eval_summary, 
    file = file.path(tables_path, "tbl1.report.summary.csv"),
    quote = TRUE, row.names = FALSE
  )
  
}

kable(
    x = eval_summary, format = report_format, row.names = FALSE,
    booktabs = TRUE, longtable = TRUE, escape = TRUE, caption = tbl_caps[1],
    format.args = list(big.mark = ","), align = c("l", "l", "r", "c", "r")
  ) %>%
  kableExtra::kable_styling(
    bootstrap_options = bootstrap_options,
    latex_options = latex_options,
    full_width = TRUE
  ) %>%
  kableExtra::column_spec(2, width = "3in")

```

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

# Specimen overview

```{r spec_summary}
spec_overview[is.na(spec_overview)] <- 0
names(spec_overview)[1] <- "Specimen"

if( args$tables ){
  
  write.csv(
    x = spec_overview, 
    file = file.path(tables_path, "tbl2.specimen.overview.csv"),
    quote = TRUE, row.names = FALSE
  )
  
}

spec_data_cols <- c(
  (ncol(spec_overview)-ifelse(umitag_option, 2, 1)):ncol(spec_overview)
)

kable(
    x = spec_overview, format = report_format, row.names = FALSE,
    booktabs = TRUE, longtable = TRUE, escape = FALSE, caption = tbl_caps[2],
    format.args = list(big.mark = ",")
  ) %>%
  kableExtra::kable_styling(
    bootstrap_options = bootstrap_options,
    latex_options = latex_options,
    full_width = TRUE
  ) %>%
  kableExtra::column_spec(column = spec_data_cols, width = "0.65in")

```

Each specimen started in the iGUIDE pipeline as genomic DNA. The gDNA was randomly sheared through ultrasonication and ligated with barcoded DNA linkers. Nested-PCR was used to amplify from incorporated dsODN sequences to the linker sequences with barcoded and linker-specific primers. This dual barcoding reduces sample to sample crossover. Amplicons were sequenced on an Illumina platform and the sequencing data processed with the iGUIDE software, available on [**GitHub@cnobles/iGUIDE**](https://github.com/cnobles/iGUIDE). 

DNA sequence reads were aligned to the `r unique(sapply(configs, "[[", "Ref_Genome"))` reference genome. The number of reads aligning for each specimen is displayed in **Table 2**, along with the number of unique "alignments" they represent (or the number of observed incorporated dsODNs). Multiple reads may represent a singular alignment of genomic DNA, inherent to sequence analysis of amplified DNA. These alignments indicate individual events of dsODN incorporation and/or clonal expansion.

Alternatively, random nucleotide sequences are included in the ligated linker sequences. These Unique Molecular Indeces (UMItags) can provide another method of abundance by counting the number of UMItags and breakpoint position combinations for each incorporation sites. This method of quantification has an increased dynamic range, yet can suffer from PCR artifacts leading to inflated abundances.

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

# On-target analysis

Incorporation sites, or locations in the genome where the dsODN was detected, are expected to be in the proximity of nuclease targeted locations. The target sequences provided for these analyses and their On-target locations (`Edit Locus`) are shown in **Table 3**. The genomic locations are in a format where chromosome, orientation, and nucleotide position are delimited by a colon (":").  

```{r target_tbl}
target_tbl$`Edit Locus` <- sapply(
  target_tbl$`Target Name`, 
  function(x) paste(on_targets[which(names(on_targets) == x)], collapse = "\n")
)

if( args$tables ){
  
  target_tbl %>%
    dplyr::mutate(
      "Edit Locus" = gsub("\n", ";", `Edit Locus`)
    ) %>%
    write.csv(
      file = file.path(tables_path, "tbl3.target.sequences.csv"),
      quote = TRUE, row.names = FALSE
    )
  
}

target_print <- target_tbl %>%
  dplyr::mutate(
    Sequence = kableExtra::cell_spec(
      Sequence, format = report_format, monospace = TRUE
    ),
    PAM = kableExtra::cell_spec(
      PAM, format = report_format, monospace = TRUE
    )
  )

if( report_format == "html" ){

  target_print <- dplyr::mutate(
    target_print,
    "Edit Locus" = gsub("\n", "<br>", `Edit Locus`)
  )
  
}else{
  
  target_print <- dplyr::mutate(
    target_print,
    "Edit Locus" = kableExtra::linebreak(`Edit Locus`, align = "c")
  )
  
}

kable(
    x = target_print, format = report_format, row.names = FALSE,
    booktabs = TRUE, longtable = TRUE, escape = FALSE, caption = tbl_caps[3],
    align = "c"
  ) %>%
  kableExtra::kable_styling(
    bootstrap_options = bootstrap_options,
    latex_options = latex_options, 
    full_width = FALSE
  )

```

Analysis of On-target associated incorporation sites (**Table 4**) produces several features that are helpful in On- and Off-target site characterization. These include the following:

* Alignment **Pileups**: unique alignments that overlap with each other or "pileup", suggesting a nearby location may be targeted for a double strand break (DSB). For this analyses, any group of `r unique(sapply(configs, "[[", "pileUpMin"))` or more unique alignments were considered as a pileup cluster.

* Flanking **Paired** alignments: alignments can be found on either side of a DSB, and therefore identifying flanking alignments suggests a DSB could be found between the paired alignments. Flanking alignments were searched for in these data up to `r 2*unique(sapply(configs, "[[", "upstreamDist"))` bp from each other.

* Target **Matched** alignments: searching for the target sequences upstream of the incorporation site can be an indicator of targeted nuclease activity. While this indicator may seem to be crucial, guide RNAs have been demonstrated to have a variety of behaviors when annealing to target DNA, not all of which can be easily searched for with a simple sequence alignment. Nucleotide sequence matching target sequences were searched for up to `r upstream_dist` bp upstream of the incorporation sites and required to have no more than `r unique(sapply(configs, "[[", "maxTargetMismatch"))` mismatches in the target sequence and/or PAM sequence.

Specimen specific tables with data relating to these criteria are found in **Table 4** for percent On-target editing and **Table `r ifelse(nrow(target_tbl) > 1, 6, 5)`** for identified Off-target loci.

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

## On-target editing efficiency

**Table 4** displays the percent of observations (efficiency or specificity) that were associated with all On-target loci for **All** alignments. Further the efficiencies for **Pileups**, **Paired**, and **Matched** criteria are displayed in the following columns. These different criteria are used as the denominator to dictcate the amount of observed nuclease-specific editing. This is an estimate though, as On-target editing does have the potential to saturate the dynamic range of the abundance calculation. Therefore, these percentages should be considered lower bounds for editing efficiency and specificity.

```{r on_target_summary}
ot_tbl_summary[is.na(ot_tbl_summary)] <- 0

if( args$tables ){
  
  ot_tbl_save <- ot_tbl_summary
  
  names(ot_tbl_save) <- c(
    "Specimen", "Condition", "Total percent", "Pileup percent", 
    "Paired percent", "Matched percent"
  )
  
  write.csv(
    x = ot_tbl_save, 
    file = file.path(tables_path, "tbl4.ontarget.summary.csv"),
    quote = TRUE, row.names = FALSE
  )
  
}

ot_tbl_summary <- ot_tbl_summary %>%
  dplyr::mutate(
    ot_algns_pct = sprintf("%.2f", round(ot_algns_pct, digits = 2)),
    ot_pile_pct = sprintf("%.2f", round(ot_pile_pct, digits = 2)),
    ot_pair_pct = sprintf("%.2f", round(ot_pair_pct, digits = 2)),
    ot_match_pct = sprintf("%.2f", round(ot_match_pct, digits = 2))
  )

ot_tbl_colnames <- c(
  "Specimen", "Condition", "Total (percent)", "Pileup (percent)", 
  "Paired (percent)", "Matched (percent)"
)

ot_tbl_alignment <- c("l", "l", rep("r", 4))

kable(
    x = ot_tbl_summary, format = report_format, digits = 4, row.names = FALSE,
    format.args = list(big.mark = ","), booktabs = TRUE, longtable = TRUE, 
    escape = FALSE, caption = tbl_caps[4], align = ot_tbl_alignment,
    col.names = ot_tbl_colnames
  ) %>%
  kableExtra::kable_styling(
    bootstrap_options = bootstrap_options,
    latex_options = latex_options,
    full_width = TRUE
  ) %>%
  kableExtra::column_spec(c(1,3:6), width = "0.65in")

```

```{r target_spec_eff, eval=any(sapply(treatment, length) > 1), child=file.path(root_dir, "tools/rscripts/report_templates/target_specific_eff_template.Rmd")}
```

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

# Off-target analysis
## Specimen information

Using the criteria discussed previously based on characterizing features of nuclease targeted sites, off-target sites can be selected from the data in an unbiased manner. **Table `r ifelse(nrow(target_tbl) > 1, 6, 5)`** shows a summary of the unique off-target locations (loci) observed in the data. For **All** alignments, the loci are based on overlapping alignments (pileup clustering) without a minimum number of fragments required to be classified as a pileup cluster. **Pileup** loci are similarly based on overlapping alignments, but require at least `r unique(sapply(configs, "[[", "pileUpMin"))` alignments to form a cluster. Flanking **Paired** loci require at least two unique alignments with opposite orientation (strands) within `r pNums(2*upstream_dist, digits = 0)` bp upstream of each other. Target **Matched** loci require a match in the upstream sequence to a treated target (within `r unique(sapply(configs, "[[", "maxTargetMismatch"))` mismatches out of the `r unique(nchar(target_tbl$Sequence))` nts`r if(any(sapply(nuc_profiles, "[[", "PAM") != FALSE)){paste0(" and ", max(sapply(nuc_profiles, function(x){if(x$PAM != FALSE){x$PAM_Tol}else{0}})), " PAM mismatch")}`).

```{r off_target_summary}
ft_tbl_summary[is.na(ft_tbl_summary)] <- 0

if( args$tables ){

  tbl_num <- ifelse(nrow(target_tbl) > 1, 6, 5)
  
  ft_tbl_save <- ft_tbl_summary
  
  names(ft_tbl_save) <- c(
    "Specimen", "Condition", "Total loci", "Pileup loci", "Paired loci", 
    "Matched loci"
  )
  
  write.csv(
    x = ft_tbl_save, 
    file = file.path(
      tables_path, paste0("tbl", tbl_num, ".offtarget.summary.csv")
    ),
    quote = TRUE, row.names = FALSE
  )
  
}


ft_tbl_colnames <- c(
  "Specimen", "Condition", 
  "Total (loci)", "Pileup (loci)", "Paired (loci)", "Matched (loci)"
)

ft_tbl_alignments <- c("l", "l", rep("r", 4))

kable(
    x = ft_tbl_summary, format = report_format, digits = 1, row.names = FALSE,
    format.args = list(big.mark = ","), booktabs = TRUE, longtable = TRUE,
    escape = FALSE, caption = tbl_caps[5], align = ft_tbl_alignments,
    col.names = ft_tbl_colnames
  ) %>%
  kableExtra::kable_styling(
    bootstrap_options = bootstrap_options,
    latex_options = latex_options,
    full_width = TRUE
  ) %>%
  kableExtra::column_spec(c(1,3:6), width = "0.65in")

```

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

## Off-target enrichment in cancer-associated genes

Flanking **Paired** loci and Target **Matched** loci are tested for enrichment against specific gene lists in **Table `r ifelse(nrow(target_tbl) > 1, 7, 6)`**. The cancer-associated and special gene lists included in this analysis were: `r paste(unique(c(sapply(configs, function(x) x$oncoGeneList$file), sapply(configs, function(x) x$specialGeneList$file))), collapse = " and ")`. Enrichment was tested by Fisher's Exact and p-values were adjusted for multiple comparisons using a Benjamani-Hochberg correction. Omitted specimens or conditions had insufficient data for this analysis (Total Gene Count = 0) or did not have enough data to support a powerful analysis (Estimated Power greater than 80%).

```{r onco_enrichment}
enrich_df <- enrich_df %>%
  dplyr::mutate(
    onco.p.value = round(onco.p.value, digits = 3),
    special.p.value = round(special.p.value, digits = 3)
  )

if( args$tables ){
  
  tbl_num <- ifelse(nrow(target_tbl) > 1, 7, 6)
  
  enrich_save <- enrich_df
  
  names(enrich_save) <- c(
    "Origin", "Condition", "Total genes", "Onco genes", "Onco p-value", 
    "Spcial genes", "Special p-value"
  )
  
  write.csv(
    x = enrich_save, 
    file = file.path(
      tables_path, paste0("tbl", tbl_num, ".gene.enrichment.csv")
    ),
    quote = TRUE, row.names = FALSE
  )
  
}

enrich_print <- enrich_df %>%
  dplyr::mutate(
    onco.power = round(onco.power * 100, digits = 0),
    onco.is.sig = onco.p.value <= 0.05,
    onco.p.value = sprintf("%1$.3f", onco.p.value),
    #onco.p.value = sprintf("%1$.3f (%2$i%%)", onco.p.value, onco.power),
    onco.p.value = kableExtra::cell_spec(
      onco.p.value, format = report_format, bold = onco.is.sig
    ),
    special.power = round(special.power * 100, digits = 0),
    special.is.sig = special.p.value <= 0.05,
    special.p.value = sprintf("%1$.3f", special.p.value),
    # special.p.value = sprintf(
    #   "%1$.3f (%2$i%%)", special.p.value, special.power
    # ),
    special.p.value = kableExtra::cell_spec(
      special.p.value, format = report_format, bold = special.is.sig
    )
  ) %>%
  dplyr::filter(
    origin == "Reference" | onco.power >= 80 | special.power >= 80
  ) %>%
  dplyr::select(-onco.is.sig, -special.is.sig, -onco.power, -special.power)

names(enrich_print) <- c(
  "Origin", "Condition", "genes", "genes", "p-val.", "genes", "p-val."
)

enrich_print[1,c(5,7)] <- ""

if( nrow(enrich_print) > 1 ){
  
  kable(
      x = enrich_print, format = report_format, row.names = FALSE,
      format.args = list(big.mark = ","), booktabs = TRUE, longtable = TRUE,
      escape = FALSE, caption = tbl_caps[6], 
      align = c("l", "l", rep("r", 5))
    ) %>%
    kableExtra::add_header_above(
      c(" " = 2, "Total", "Onco Enrich." = 2, "Special Enrich." = 2)
    ) %>%
    kableExtra::kable_styling(
      bootstrap_options = bootstrap_options,
      latex_options = latex_options,
      full_width = TRUE
    ) %>%
    kableExtra::column_spec(1, width = "1.1in") %>%
    kableExtra::column_spec(2, width = "3in")
  
}else{
  cat("**All comparisons were omitted due to lack of sufficient data.**")
}

```

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

## Genomic distribution of incorporation sites

The figure(s) below display the genomic distribution of identified incorporation sites. The inner most ring plots all alignments identified within the associated data, while subsequent rings plot the alignments associated with Pileups, Flanking Pairs, and Target Matched groups. The height of the bar within its associated ring is correlated to the number of incorporations identified within the **10 Mb window** (logarithm base 10 of incorporation site abundances).

```{r genomic_distribution}
genomic_den_plot <- plotGenomicDensity(
    genomic_grl, 
    res = 1E7, 
    grp.col = "condition", 
    abund.method = "log.count"
  ) +
  custom_theme +
  theme(
    panel.background = element_blank(),
    panel.border = element_blank(),
    axis.ticks = element_blank(), 
    legend.position = "top"
  )

gen_den_len <- length(unique(genomic_den_plot$data$cond))
plots_per_row <- 2
row_idx <- ceiling(seq_len(gen_den_len) / plots_per_row)
layout_list <- split(seq_len(gen_den_len), row_idx)

null <- lapply(
  seq_along(layout_list),
  function(i, gd_plot, fp){
    
    p <- gd_plot + 
      facet_wrap_paginate(~ cond, ncol = 2, nrow = 1, page = i)

    file_pdf <- sprintf("genomic_dens-%s.pdf", i)
    file_png <- sprintf("genomic_dens-%s.png", i)
    
    ggsave(
      filename = file_pdf, 
      plot = p, 
      device = "pdf", 
      path = fp, 
      width = figure_width, 
      height = 4.7, 
      units = "in"
    )
    
    ggsave(
      filename = file_png, 
      plot = p, 
      device = "png", 
      path = fp, 
      width = figure_width,
      height = 4.7, 
      units = "in",
      dpi = knitr::opts_chunk$get("dpi")
    )

  },
  gd_plot = genomic_den_plot, 
  fp = figure_path
)

if( args$format == "pdf"){
  
  knitr::include_graphics(
    path = sprintf(
      file.path(figure_path, "genomic_dens-%s.pdf"), 
      seq_along(layout_list)
    )
  )
  
}else{
  
  knitr::include_graphics(
    path = sprintf(
      file.path(figure_path, "genomic_dens-%s.png"), 
      seq_along(layout_list)
    )
  )
  
}
```

```{r} 
cat(fig_caps[1]) 
```

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

## Off-target sequence comparison

Off-target sites can be identified by sequence similarity within `r format(upstream_dist, big.mark = ",")` bp upstream of incorporation sites. The sequences of the top target matched sites are displayed below in Figure 3 for each condition and target along with the number of mismatches to the targeting sequence (`Mismatch`), an indication if the site is associated with an On- or Off-target location (`Target`), the total number of unique alignments associated with the site (`Abund.`), the maximum edit site likelyhood (`MESL`), and an identifier denoted by the nearest gene (`Gene_ID`). MESL is a score for the percentage likelyhood the off-target site is associated with directed nuclease editing, based solely on the respective On-target incorporation distribution. The gene name within the `Gene_ID` is the nearest gene to the genomic location. Further, symbols after the gene name indicate: `*` that the site is within the transcription unit of the gene, `~` the gene appears on the cancer-association list, `!` and that the gene appears on the special gene list. For this report, gene lists used were: `r paste(unique(c(sapply(configs, function(x) x$oncoGeneList$file), sapply(configs, function(x) x$specialGeneList$file))), collapse = " and ")`. For a complete list of predicted Off-target sites identified by the analysis, please see the Supporting material section, Off-target site tables.

```{r off_target_seqs}
ft_seq_plots <- lapply(
  ft_seqs_list, 
  function(x){
    
    ref_target_seqs <- sapply(seq_len(nrow(target_tbl)), function(i){
      
      prof <- nuc_profiles[[target_tbl$Nuclease[i]]]
      
      if( prof$PAM_Loc == "5p" ){
        return(paste0(target_tbl$PAM[i], target_tbl$Sequence[i]))
      }else if( prof$PAM_Loc == "3p" ){
        return(paste0(target_tbl$Sequence[i], target_tbl$PAM[i]))
      }else{
        return(target_tbl$Sequence[i])
      }
      
    })
    
    ref_seq <- ref_target_seqs[match(unique(x$target.seq), target_tbl$Target)]
    
    if( support_present ){
      title <- paste0(unique(x$condition), "\nTarget: ", unique(x$target.seq))
    }else{
      title <- paste0("Target: ", unique(x$target.seq))
    }
    
    y <- head(x, n = 10)
    
    dplyr::select(
        y, aligned.sequence, mismatch, target, aligns, MESL, gene_id
      ) %>%
      dplyr::rename(
        "Mismatch" = mismatch, "Target" = target, 
        "Abund." = aligns, "Gene_ID" = gene_id
      ) %>%
      dplyr::mutate(MESL = sprintf("%.1f", round(MESL, digits = 1))) %>%
      plotSeqDiverge(
        ref = ref_seq, padding = 7, text.size = 3.5, 
        font.family = "Courier", font.face = "plain", force.sq = TRUE) +
      labs(title = title) +
      blank_theme +
      theme(
        title = element_text(size = 9, face = "bold")
      )
    
  }
)

layout_seq <- c(
  rep(seq_along(ft_seqs_list), sapply(
    ft_seqs_list, 
    function(x) min(nrow(x), 10)
  ) + 4)
)

lines_per_page <- 42
layout_list <- split(layout_seq, layout_seq)

null <- mapply(
  function(seq, ft_plot, fp){
    
    len <- length(seq) * 10 / lines_per_page
    file_pdf <- sprintf("off_target_seqs-%s.pdf", unique(seq))
    file_png <- sprintf("off_target_seqs-%s.png", unique(seq))
    
    ggsave(
      filename = file_pdf, 
      plot = ft_plot, 
      device = "pdf", 
      path = fp, 
      width = 7.5, 
      height = len, 
      units = "in"
    )
    
    ggsave(
      filename = file_png, 
      plot = ft_plot, 
      device = "png", 
      path = fp, 
      width = 7.5,
      height = len, 
      units = "in",
      dpi = 300
    )

    
  },
  seq = layout_list,
  ft_plot = ft_seq_plots,
  MoreArgs = list(fp = figure_path)
)

if( args$format == "pdf"){
  
  knitr::include_graphics(
    path = sprintf(
      file.path(figure_path, "off_target_seqs-%s.pdf"), 
      seq_along(layout_list)
    )
  )
  
  
}else{
  
  knitr::include_graphics(
    path = sprintf(
      file.path(figure_path, "off_target_seqs-%s.png"), 
      seq_along(layout_list)
    )
  )
  
}

```

```{r} 
cat(fig_caps[2]) 
```

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

# Supporting material

## Editing distributions at On-target loci

**Figure 3** displays the distribution of dsODN incorporations around On-target site(s). Incorporations in different orientations are shown on the positive (red) and negative (blue) y-axis. The percentage in the bottom right corner of each plot is an estimate of the number of incorporations associated with the on-target site (based on pileups) captured within the allowed window of `r unique(sapply(configs, "[[", "upstreamDist"))` bps. These data can be used to fine tune the processing analyses, specifically the `upstreamDist` parameter which modifies the distance upstream of incorporation sites to search for nuclease edited sequences. 

```{r}
incorp_len <- ifelse(
  nrow(eval_data$spec_info$supp_data) > 0,
  length(unique(paste(on_tar_dists$condition, on_tar_dists$target))), 
  length(unique(on_tar_dists$target))
)
```

```{r incorp_dist}
incorp_plot <- ggplot(on_tar_dists, aes(x = edit.site.dist, y = strand.cnt)) +
  geom_vline(xintercept = 0, color = "black", linetype = "dotted") +
  geom_col(aes(fill = factor(strand)), width = 1) +
  geom_text(
    data = sites_included, 
    aes(x = x_pos, y = y_pos, label = prop), 
    hjust = 1, fontface = "bold", size = 5) +
  coord_cartesian(xlim = c(-upstream_dist, upstream_dist)) +
  scale_y_continuous(breaks = pretty_breaks(), labels = pNums) +
  scale_fill_brewer(type = "qual", palette = "Set1") + 
  guides(fill = FALSE) +
  labs(
    x = "Distance to Edit Site (bp, res = 1)", 
    y = "Log Alignment Count") +
  custom_theme +
  theme(
    strip.text.y = element_text(angle = 0),
    aspect.ratio = 7/12
  )


null <- lapply(
  unique(ceiling(seq_len(incorp_len) / 2)),
  function(i, ic_plot, fp, supp_present){
    
    if( supp_present ){
  
      p <- ic_plot + 
        facet_wrap_paginate(
          condition ~ target, ncol = 2, nrow = 1, page = i, scales = "free"
        )
    
    }else{
    
      p <- ic_plot + 
        facet_wrap_paginate(
          ~ target, ncol = 2, nrow = 1, page = i, scales = "free"
        )
      
    }
    
    
    file_pdf <- sprintf("incorp_dist-%s.pdf", i)
    file_png <- sprintf("incorp_dist-%s.png", i)
    
    ggsave(
      filename = file_pdf, 
      plot = p, 
      device = "pdf", 
      path = fp, 
      width = figure_width, 
      height = 3, 
      units = "in"
    )
    
    ggsave(
      filename = file_png, 
      plot = p, 
      device = "png", 
      path = fp, 
      width = figure_width,
      height = 3, 
      units = "in",
      dpi = knitr::opts_chunk$get("dpi")
    )

  },
  ic_plot = incorp_plot, 
  fp = figure_path,
  supp_present = nrow(eval_data$spec_info$supp_data) > 0
)

if( args$format == "pdf"){
  
  knitr::include_graphics(
    path = sprintf(
      file.path(figure_path, "incorp_dist-%s.pdf"), 
      unique(ceiling(seq_len(incorp_len) / 2))
    )
  )
  
}else{
  
  knitr::include_graphics(
    path = sprintf(
      file.path(figure_path, "incorp_dist-%s.png"), 
      unique(ceiling(seq_len(incorp_len) / 2))
    )
  )
  
}
  
```

```{r} 
cat(fig_caps[3]) 
```

```{r} 
if( args$format == "pdf" ){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

## Off-target site tables

Below is a complete list of predicted Off-target sites identified for each condition and target. The additional column `Edit Site` (compared to Figure 3) indicates the predicted genomic editing location, and is broken into three parts delimited by ":". The first part indicates the chromosome, the second the orientation with respect to genomic numbering (sense "+" or anti-sense "-"), and the third is the location of the targeted base pair. 

```{r off_target_tbls}

full_target_seqs <- structure(
  sapply(seq_len(nrow(target_tbl)), function(i){
    
    nuc <- target_tbl$Nuclease[i]
    sequence <- target_tbl$Sequence[i]

    ifelse(
      nuc_profiles[[nuc]]$PAM_Loc == "3p",
      paste0(sequence, nuc_profiles[[nuc]]$PAM),
      ifelse(
        nuc_profiles[[nuc]]$PAM_Loc == "5p",
        paste0(nuc_profiles[[nuc]]$PAM, sequence),
        sequence
      )
    )
  
  }),
  names = target_tbl$`Target Name`
)

last_tbl_num <- ifelse(nrow(target_tbl) > 1, 7, 6)

for( i in seq_along(ft_seqs_list) ){
  
  cap <- paste0("Off-Target Loci: ", names(ft_seqs_list)[i])
  
  if( report_format == "html" ){
    cap <- paste0("Table ", i + last_tbl_num, ". ", cap)
  }
  
  target_ref_seq <- full_target_seqs[unique(ft_seqs_list[[i]]$target.seq)]
  
  df <- dplyr::select(
      ft_seqs_list[[i]], target, gene_id, edit.site, aligns, MESL, 
      aligned.sequence, mismatch
    ) %>%
    dplyr::mutate(
      aligned.sequence = divSeq(
        seqs = aligned.sequence, 
        ref = target_ref_seq
      ),
      MESL = round(MESL, digits = 1)
    )
  
  if( args$tables ){
    
    df %>%
      dplyr::rename(
        "Target" = target, "Gene ID" = gene_id, "Edit Site" = edit.site, 
        "Abund." = aligns, "Aligned Sequence" = aligned.sequence, 
        "Mismatch" = mismatch
      )%>%
      write.csv(
        file = file.path(
          tables_path, paste0("tbl", i + last_tbl_num, ".offtargets.csv")
        ),
        quote = TRUE, row.names = FALSE
      )
    
  }

  dplyr::mutate(
      df,
      MESL = sprintf("%.1f", MESL),
      aligned.sequence = ifelse(
        rep(report_format, n()) == rep("html", n()),
        gsub(".", "\\.", aligned.sequence, fixed = TRUE),
        aligned.sequence
      ),
      aligned.sequence = kableExtra::cell_spec(
        aligned.sequence, format = report_format, monospace = TRUE, 
        escape = FALSE
      ),
      gene_id = ifelse(
        rep(report_format, n()) == rep("html", n()),
        gene_id,
        gsub("_", "\\_", gene_id, fixed = TRUE)
      ),
      gene_id = ifelse(
        rep(report_format, n()) == rep("html", n()),
        gene_id,
        gsub("~", "\\~", gene_id, fixed = TRUE)
      ),
      gene_id = ifelse(
        rep(report_format, n()) == rep("html", n()),
        gene_id,
        gsub("!", "\\!", gene_id, fixed = TRUE)
      ),
      gene_id = kableExtra::cell_spec(
        gene_id, format = report_format, escape = FALSE
      )
    ) %>%
    dplyr::rename(
     "Target" = target, "Gene ID" = gene_id, "Edit Site" = edit.site, 
     "Abund." = aligns, "Aligned Sequence" = aligned.sequence, 
     "Mismatch" = mismatch
    ) %>%
    kable(
      format = report_format, row.names = FALSE,
      format.args = list(big.mark = ","), booktabs = TRUE, longtable = TRUE,
      escape = FALSE, align = c(rep("c", 3), "r", "r", "c", "c"), caption = cap
    ) %>%
    kableExtra::kable_styling(
      bootstrap_options = bootstrap_options,
      latex_options = latex_options
    ) %>%
    kableExtra::column_spec(c(1, 4, 5), width = "0.4in") %>%
    kableExtra::column_spec(c(7), width = "0.55in") %>%
    kableExtra::column_spec(c(2, 3), width = "1.25in") %>%
    kableExtra::column_spec(c(6), width = "2in") %>%
    print()
  
}

```
